apollo-router/src/services/supergraph/service.rs

```diff
@@ -612,19 +613,40 @@ async fn plan_query(
         context.extensions().lock().insert::<ParsedDocument>(doc);
     }
 
-    planning
+    let qpr = planning
         .call(
             query_planner::CachingRequest::builder()
                 .query(query_str)
                 .and_operation_name(operation_name)
-                .context(context)
+                .context(context.clone())
                 .build(),
         )
         .instrument(tracing::info_span!(
             QUERY_PLANNING_SPAN_NAME,
             "otel.kind" = "INTERNAL"
         ))
-        .await
+        .await?;
+
+    // We need to operate on the BatchQuery without holding on to the extension lock, but using the
+    // batched query requires awaiting it, which would hold the lock for too long.
+    //
+    // We remove the BatchQuery here and then reinsert it after we've done our call to fix that.
+    let batching = context.extensions().lock().remove::<BatchQuery>();
+    if let Some(mut batch_query) = batching {
+        if let Some(QueryPlannerContent::Plan { plan, .. }) = &qpr.content {
+            let mut query_hashes = vec![];
+            plan.root.query_hashes(&mut query_hashes);
+            batch_query
+                .set_query_hashes(query_hashes)
+                .await
+                .map_err(|e| CacheResolverError::BatchingError(e.to_string()))?;
+            tracing::debug!("batch registered: {}", batch_query);
+        }
+
+        context.extensions().lock().insert(batch_query);
+    }
```

Comment:
I know what you mean. It makes me nervous too. I *believe* that there are no races as currently written, since nothing should be interacting with this extension when this process is performed.
I suppose it would be possible to cause issue if you've spawned a long running task in RouterRequest processing that is monitoring the value of this extension. If you are doing things like that, then you probably have a lot of issues.

I could restructure this so that we don't make the call to `set_query_hashes()` under the lock. Something like:

```
    let mut query_hashes = vec![];
    if let Some(mut batch_query) = context.extensions().lock().get_mut::<BatchQuery>() {
        if let Some(QueryPlannerContent::Plan { plan, .. }) = &qpr.content {
            plan.root.query_hashes(&mut query_hashes);
        }
    }
    if !query_hashes.is_empty() {
            batch_query
                .set_query_hashes(query_hashes)
                .await
                .map_err(|e| CacheResolverError::BatchingError(e.to_string()))?;
            tracing::debug!("batch registered: {}", batch_query);
    }
```

Reply:
I've raised a PR against this PR to explore this: https://github.com/apollographql/router/pull/4941

Please let me know what you think about that.

Reply:
done: 86855eebe883f211274a51d9449eeea5e51a0aaf